### ISSUE TITLE
A/F: `getSignificantBits() <= 64 && "Too many bits for int64_t"' after 6c951c5ee6d0b

### ISSUE THREAD

**Issue by dyung** [llvm/llvm-project#56170]
We have an internal test that started hitting an assertion failure which I bisected to commit 6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5. Consider the following code:
```cpp
enum a { b, c };
enum d { e, f };
struct g {
  unsigned h : 16;
  unsigned i : 16;
  unsigned j : 16;
  unsigned ab : 8;
  unsigned : 8;
  unsigned ac : 6;
  d k : 6;
  void l(g, int *);
} m;
int n;
void o() {
  struct p {
    int q;
  };
  struct r {
    int s;
    int i;
    p t;
  } ad = *(r *)o;
  a ah = ad.t.q ? c : b;
  g aj;
  aj.h = aj.i = ad.i;
  aj.j = ad.s;
  aj.ab = aj.ac = ah;
  aj.k = f;
  m.l(aj, &n);
}
```
If compiled with optimizations enabled with a compiler built that includes 6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5, you will hit an assertion failure:
```console
clang: /home/dyung/src/upstream/llvm_clean_git/llvm/include/llvm/ADT/APInt.h:1481: int64_t llvm::APInt::getSExtValue() const: Assertion `getSignificantBits() <= 64 && "Too many bits for int64_t"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang -c repro.cpp -O2
1.	<eof> parser at end of file
2.	Code generation
3.	Running pass 'Function Pass Manager' on module 'repro.cpp'.
4.	Running pass 'X86 DAG->DAG Instruction Selection' on function '@_Z1ov'
 #0 0x000055894514d264 PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #1 0x000055894514afac llvm::sys::CleanupOnSignal(unsigned long) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x3da6fac)
 #2 0x00005589450835b8 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x00007f972c207420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #4 0x00007f972bca400b raise /build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #5 0x00007f972bc83859 abort /build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:81:7
 #6 0x00007f972bc83729 get_sysdep_segment_value /build/glibc-SzIz7B/glibc-2.31/intl/loadmsgcat.c:509:8
 #7 0x00007f972bc83729 _nl_load_domain /build/glibc-SzIz7B/glibc-2.31/intl/loadmsgcat.c:970:34
 #8 0x00007f972bc94fd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #9 0x000055894614c5f7 (anonymous namespace)::DAGCombiner::visitADDLike(llvm::SDNode*) DAGCombiner.cpp:0:0
#10 0x000055894614c8ac (anonymous namespace)::DAGCombiner::visitADD(llvm::SDNode*) DAGCombiner.cpp:0:0
#11 0x0000558946182f7d (anonymous namespace)::DAGCombiner::visit(llvm::SDNode*) DAGCombiner.cpp:0:0
#12 0x00005589461849b5 (anonymous namespace)::DAGCombiner::combine(llvm::SDNode*) DAGCombiner.cpp:0:0
#13 0x0000558946186240 llvm::SelectionDAG::Combine(llvm::CombineLevel, llvm::AAResults*, llvm::CodeGenOpt::Level) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x4de2240)
#14 0x00005589462a08c8 llvm::SelectionDAGISel::CodeGenAndEmitDAG() (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x4efc8c8)
#15 0x00005589462a4973 llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x4f00973)
#16 0x00005589462a6acc llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) (.part.0) SelectionDAGISel.cpp:0:0
#17 0x0000558943ad21e0 (anonymous namespace)::X86DAGToDAGISel::runOnMachineFunction(llvm::MachineFunction&) X86ISelDAGToDAG.cpp:0:0
#18 0x0000558944289438 llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x2ee5438)
#19 0x00005589447c2b25 llvm::FPPassManager::runOnFunction(llvm::Function&) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x341eb25)
#20 0x00005589447c2d69 llvm::FPPassManager::runOnModule(llvm::Module&) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x341ed69)
#21 0x00005589447c35b2 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x341f5b2)
#22 0x000055894550f972 clang::EmitBackendOutput(clang::DiagnosticsEngine&, clang::HeaderSearchOptions const&, clang::CodeGenOptions const&, clang::TargetOptions const&, clang::LangOptions const&, llvm::StringRef, llvm::Module*, clang::BackendAction, std::unique_ptr<llvm::raw_pwrite_stream, std::default_delete<llvm::raw_pwrite_stream>>) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x416b972)
#23 0x00005589463f5a0b clang::BackendConsumer::HandleTranslationUnit(clang::ASTContext&) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x5051a0b)
#24 0x0000558947235169 clang::ParseAST(clang::Sema&, bool, bool) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x5e91169)
#25 0x00005589463f4498 clang::CodeGenAction::ExecuteAction() (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x5050498)
#26 0x0000558945ca0cd9 clang::FrontendAction::Execute() (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x48fccd9)
#27 0x0000558945c2bcfe clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x4887cfe)
#28 0x0000558945d7fec3 clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x49dbec3)
#29 0x00005589426937c1 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x12ef7c1)
#30 0x000055894268c8f8 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) driver.cpp:0:0
#31 0x0000558945a9edf9 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const::'lambda'()>(long) Job.cpp:0:0
#32 0x000055894508373c llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x3cdf73c)
#33 0x0000558945a9f50f clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef>>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>*, bool*) const (.part.0) Job.cpp:0:0
#34 0x0000558945a69f39 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&, bool) const (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x46c5f39)
#35 0x0000558945a6aa5d clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&, bool) const (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x46c6a5d)
#36 0x0000558945a74b0c clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*>>&) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x46d0b0c)
#37 0x0000558942691b83 clang_main(int, char**) (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x12edb83)
#38 0x00007f972bc85083 __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:342:3
#39 0x000055894268c4ee _start (/home/dyung/src/upstream/6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5-linux/bin/clang+0x12e84ee)
clang-15: error: clang frontend command failed with exit code 134 (use -v to see invocation)
clang version 15.0.0 (https://github.com/llvm/llvm-project.git 6c951c5ee6d0b848877cb8ac7a9cb2a9ef9ebbb5)
```

**Comment by RKSimon** [comment]
reduced:
```
target triple = "x86_64-unknown-linux-gnu"

define dso_local void @_Z1ov() {
entry:
  %ad.sroa.5.0.copyload = load i32, ptr getelementptr inbounds (i8, ptr @_Z1ov, i64 8), align 4
  %tobool.not.not = icmp eq i32 %ad.sroa.5.0.copyload, 0
  %bf.shl12 = select i1 %tobool.not.not, i80 0, i80 18446744073709551616
  %bf.shl18 = select i1 %tobool.not.not, i80 0, i80 281474976710656
  %bf.set9 = or i80 0, %bf.shl12
  %bf.set14 = or i80 %bf.set9, 1180591620717411303424
  %bf.clear22 = add nuw nsw i80 %bf.set14, %bf.shl18
  store i80 %bf.clear22, ptr undef, align 8
  ret void
}
```

**Comment by vfdff** [comment]
it doesn't reproduce base on the new upstream, https://gcc.godbolt.org/z/dshc7Wchj

**Comment by bcl5980** [comment]
fixed by 8c742056
